### PR TITLE
✨ feat(editable): update for display panel modal

### PR DESCRIPTION
**Pull request by mehdiezz21**
Update Editable for display panel modal.

https://github.com/p3ol/oak/assets/72040763/aaa64a38-f7ee-41a1-9a75-70e0671d75de

### CODE REVIEW

**Comment by maximedasilva** [comment]
may be I'm wrong, but I thought the editable type was an option passed to the builder to allow us to use floating or modal when defining a new oak Builder

**Comment by dackmin** [comment]
> Ugo Stephant 24/06/2024 15h06
editableType: 'floating' (default) | 'panel'

**Review by dackmin — CHANGES_REQUESTED** [review]
`editableType` needs to be a `BuilderOption` available on both core & the react renderer, it's retrieved from `useBuilder` here but doesn't come from anywhere. 
Add it to `packages/core/lib/classes.ts:BuilderOptions` & `packages/core/lib/types.ts:BuilderObject` and start from here.

**Comment by dackmin** [comment]
can you add a story for the modal type?

File: packages/react/lib/components/Col/index.tsx [hunk 1]
```diff
@@ -35,6 +36,7 @@ const Col = ({
   onRemove,
 }: ColProps) => {
   const editableRef = useRef<EditableRef>();
+  const modalRef: MutableRefObject<ModalRef> = useRef();
```
Comment (dackmin) [inline packages/react/lib/components/Col/index.tsx:39]: `const modalRef = useRef<ModalRef>();`

File: packages/react/lib/Element/index.tsx [hunk 2]
```diff
@@ -53,6 +55,7 @@ const Element = forwardRef<ElementRef, ElementProps>(({
 }, ref) => {
   const innerRef = useRef<DroppableRef>();
   const editableRef = useRef<EditableRef>();
+  const modalRef: MutableRefObject<ModalRef> = useRef();
```
Comment (dackmin) [inline packages/react/lib/Element/index.tsx:58]: `const modalRef = useRef<ModalRef>();`

File: packages/react/lib/Editable/Modal.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import { useBuilder } from '../hooks';
+
+export interface ModalEditableProps extends SpecialComponentPropsWithoutRef {
+  children: ReactElement;
+  element: ElementObject;
+  component: ComponentObject;
+  modalRef?: MutableRefObject<any>;
+}
+
+export declare interface FloatingRef extends OakRef {
```
Comment (dackmin) [inline packages/react/lib/Editable/Modal.tsx:30]: remove this, I don't think it's used inside Modal

File: packages/react/lib/Editable/Modal.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import {
+  Modal,
+  type SpecialComponentPropsWithoutRef,
+} from '@junipero/react';
+import { slideInLeftModal } from '@junipero/transitions';
+
+import type { EditableRef } from '.';
+import type { OakRef } from '../types';
+import Form from './Form';
+import { useBuilder } from '../hooks';
```
Comment (dackmin) [inline packages/react/lib/Editable/Modal.tsx:21]: move before form

File: packages/react/lib/Editable/index.tsx [hunk 5]
```diff
@@ -1,70 +1,51 @@
...
-import { useBuilder } from '../hooks';
-import Form from './Form';
+import FloatingEditable from './Floating';
+import ModalEditable from './Modal';
 
 export interface EditableProps extends SpecialComponentPropsWithoutRef {
   children: ReactElement;
   element: ElementObject;
   component: ComponentObject;
+  modalRef?: MutableRefObject<ModalRef>;
```
Comment (dackmin) [inline packages/react/lib/Editable/index.tsx:24]: ModalRef is already a MutableRefObject, you can just use ModalRef

File: packages/react/lib/Editable/index.tsx [hunk 6]
```diff
@@ -1,70 +1,51 @@
...
+  opened?: boolean;
 }
 
 export declare interface FloatingRef extends OakRef {
-  open: () => void;
-  close: () => void;
-  toggle: () => void;
-  opened: boolean;
-  innerRef: MutableRefObject<any>;
+  open?: () => void;
```
Comment (dackmin) [inline packages/react/lib/Editable/index.tsx:34]: FloatingRef is not used here, remove it

File: packages/react/lib/Editable/index.tsx [hunk 7]
```diff
@@ -1,70 +1,51 @@
...
-  close: () => void;
-  forceClose: () => void;
-  toggle: () => void;
-  innerRef: MutableRefObject<FloatingRef | HTMLDivElement>;
+  open?: () => void;
+  close?: () => void;
+  toggle?: () => void;
+  forceClose?: () => void;
+  innerRef?: MutableRefObject<FloatingRef | HTMLDivElement | null>;
+  current?: FloatingRef | HTMLDivElement | null;
```
Comment (dackmin) [inline packages/react/lib/Editable/index.tsx:48]: current does not exist

**Review by dackmin — CHANGES_REQUESTED** [review]

**Review by dackmin — APPROVED** [review]

File: packages/react/lib/Editable/Floating.tsx [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+  type MouseEvent,
+  Children,
+  cloneElement,
+  useImperativeHandle,
+  useMemo,
+  useReducer,
+  useRef,
+  forwardRef,
+} from 'react';
+import type {
```
Comment (maximedasilva) [inline packages/react/lib/Editable/Floating.tsx:13]: import full type first

File: packages/react/lib/Editable/Modal.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
+import {
+  type MutableRefObject,
+  type ReactElement,
+  type MouseEvent,
+  cloneElement,
+  forwardRef,
+} from 'react';
+import type {
```
Comment (maximedasilva) [inline packages/react/lib/Editable/Modal.tsx:8]: import full type first

File: packages/react/lib/Editable/index.tsx [hunk 10]
```diff
@@ -1,70 +1,41 @@
...
-  Children,
-  cloneElement,
   forwardRef,
-  useImperativeHandle,
-  useMemo,
-  useReducer,
-  useRef,
 } from 'react';
 import type {
   ComponentObject,
```
Comment (maximedasilva) [inline packages/react/lib/Editable/index.tsx:7]: import full type first

File: packages/react/lib/Editable/index.tsx [hunk 11]
```diff
@@ -1,70 +1,41 @@
...
-  ComponentSettingsFormObject,
   ElementObject,
 } from '@oakjs/core';
-import { createPortal } from 'react-dom';
 import {
```
Comment (maximedasilva) [inline packages/react/lib/Editable/index.tsx:10]: its a full type import

File: packages/react/lib/components/Col/index.tsx [hunk 12]
```diff
@@ -2,11 +2,12 @@ import {
   type ComponentPropsWithoutRef,
   type Key,
   type MouseEvent,
+  type MutableRefObject,
   useMemo,
   useRef,
 } from 'react';
 import type { ComponentObject, ComponentOverrideObject, ElementObject } from '@oakjs/core';
```
Comment (maximedasilva) [inline packages/react/lib/components/Col/index.tsx:9]: import full type first

**Review by maximedasilva — CHANGES_REQUESTED** [review]

**Pull request merged**